### PR TITLE
No CORS headers sent if Exception is thrown

### PR DESCRIPTION
**Pull request by pranavkm**
- Normally headers are added however if a controller throws an exception then CORS headers not be present.

Addresses aspnet/Home#3220

### CODE REVIEW

**Comment by pranavkm** [comment]
Cleaned up formatting + LoggerExtensions

File: src/Microsoft.AspNetCore.Cors/Infrastructure/LoggerExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+{
+    internal static class LoggerExtensions
+    {
+        private static readonly Action<ILogger, Exception> _failedToSetCorsHeaders;
+
+        static LoggerExtensions()
+        {
+            _failedToSetCorsHeaders = LoggerMessage.Define(
+                LogLevel.Warning,
+                1,
```
Comment (kichalla) [inline src/Microsoft.AspNetCore.Cors/Infrastructure/LoggerExtensions.cs]: There should be other logging events from Cors, this new one should be after them

File: test/Microsoft.AspNetCore.Cors.Test/CorsMiddlewareTests.cs [hunk 2]
```diff
@@ -349,5 +354,39 @@ public class CorsMiddlewareTests
...
+        {
+            // Arrange
+            var hostBuilder = new WebHostBuilder()
+                .Configure(app => {
+                    app.UseCors(builder =>
+                        builder.WithOrigins("http://localhost:5001")
+                            .WithMethods("PUT")
+                            .WithHeaders("Header1")
+                            .WithExposedHeaders("AllowedHeader"));
+                    app.Run(async context => {
```
Comment (kichalla) [inline test/Microsoft.AspNetCore.Cors.Test/CorsMiddlewareTests.cs]: Would be good to mimic a similar scenario as the original issue (i.e to register a middleware _before_ Cors and which catches exceptions globally and clears the response headers and sends out a generic safe response)

File: src/Microsoft.AspNetCore.Cors/Infrastructure/CorsMiddleware.cs [hunk 3]
```diff
@@ -40,10 +44,61 @@ public class CorsMiddleware
...
+        {
+        }
+
+        /// <summary>
+        /// Instantiates a new <see cref="CorsMiddleware"/>.
+        /// </summary>
+        /// <param name="next">The next middleware in the pipeline.</param>
+        /// <param name="corsService">An instance of <see cref="ICorsService"/>.</param>
+        /// <param name="policy">An instance of the <see cref="CorsPolicy"/> which can be applied.</param>
+        [Obsolete("This constructor has been replaced with an equivalent constructor which requires an ILoggerFactory")]
```
Comment (kichalla) [inline src/Microsoft.AspNetCore.Cors/Infrastructure/CorsMiddleware.cs]: Remove this constructor?

Comment (pranavkm) [reply src/Microsoft.AspNetCore.Cors/Infrastructure/CorsMiddleware.cs]: GitHub is doing a bad job doing a diff. There were 3 original ctos - https://github.com/aspnet/CORS/blob/dev/src/Microsoft.AspNetCore.Cors/Infrastructure/CorsMiddleware.cs - that now have an overload each.

File: src/Microsoft.AspNetCore.Cors/Infrastructure/LoggerExtensions.cs (hunk 1 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Cors/Infrastructure/LoggerExtensions.cs]: Ahh, didn't notice it was sitting under the Internal ns.

File: test/Microsoft.AspNetCore.Cors.Test/CorsMiddlewareTests.cs [hunk 4]
```diff
@@ -349,5 +355,121 @@ public class CorsMiddlewareTests
                 Assert.Equal("PUT", response.Headers.GetValues(CorsConstants.AccessControlAllowMethods).FirstOrDefault());
             }
         }
+
+        [Fact]
+        public async Task CorsRequest_AfterResponseCleared_SetsResponseHeaders()
```
Comment (kichalla) [inline test/Microsoft.AspNetCore.Cors.Test/CorsMiddlewareTests.cs]: nit: change name

**Pull request merged**